gstd/src/common/reservations.rs

```diff
@@ -0,0 +1,328 @@
+///             .expect("failed to reserve gas");
+///         RESERVATIONS
+///             .reserve(50_000, 30)
+///             .expect("failed to reserve gas");
+///     }
+/// }
+///
+/// extern "C" fn handle() {
+///     let reservation = unsafe { RESERVATIONS.try_take_reservation(100_000) };
+///     match reservation {
+///         Some(id) => {
+///             msg::send_bytes_from_reservation(
+///                 id,
+///                 msg::source(),
+///                 b"send_bytes_from_reservation",
+///                 0,
+///             )
+///             .expect("Failed to send message from reservation");
+///         }
+///         None => {
+///             msg::send_bytes(msg::source(), b"send_bytes", 0).expect("Failed to send message");
+///         }
+///     }
+/// }
+/// ```
+///
+/// # See also
+/// - [`ReservationId`](ReservationId) is used to reserve and unreserve gas
+/// for program execution later.
+#[derive(Default)]
+pub struct Reservations(Vec<Reservation>);
+
+impl Reservations {
+    /// Create a new [`Reservations`] struct.
+    pub const fn new() -> Self {
+        Reservations(Vec::new())
+    }
+
+    /// Reserve the `amount` of gas for further usage.
+    ///
+    /// `duration` is the block count within which the reservation must be used.
+    ///
+    /// # Underlying logics
+    ///
+    /// Executes for O(logN)..O(N), where N is a number of stored
+    /// reservations.
+    ///
+    /// All the reservations are kept sorted by amount in
+    /// ascending order(when amount is the same, they're sorted by time when
+    /// they expire) when inserted, so the closer inserted element to the
+    /// beginning of the underlying `Vec` the closer execution time will be
+    /// to O(N).
+    ///
+    /// Also, when the underlying `Vec` will allocate new memory
+    /// the attempt to clean expired reservations occurs to avoid memory
+    /// allocations.
+    pub fn reserve(&mut self, amount: u64, duration: u32) -> Result<()> {
+        let id = ReservationId::reserve(amount, duration)?;
+        let block_height = exec::block_height();
+        let valid_until = duration.saturating_add(block_height);
+
+        let insert_range_start = self
+            .0
+            .partition_point(|reservation| reservation.amount < amount);
+        let insert_range_end = self
+            .0
+            .partition_point(|reservation| reservation.amount <= amount);
+
+        let insert_to = self.0[insert_range_start..insert_range_end]
+            .binary_search_by(|reservation| reservation.valid_until.cmp(&valid_until));
+        let insert_to = if insert_range_start == self.0.len() {
+            self.0.len()
+        } else {
+            match insert_to {
+                Ok(pos) => pos + 1,
+                Err(pos) => pos,
+            }
+        };
+
+        // self.0 will allocate new memory.
+        if self.0.capacity() == self.0.len() {
+            self.cleanup(block_height);
```

Comment:
Any cleanup should occur only in-place when you query some reservation and looking for it using bsearch or manually

Reply:
Or by schedule: e.g. each 500 blocks on first interaction with manager

Reply:
well, current implementation actually performs on-query invalidation for the range of values in `unreserve`.

As invalidation done on the, imho, right level (during `unreserve`), I don't see any problems with including it in the `reserve`. What's more, it safes from unneeded allocations.